### ISSUE TITLE
policy/v1beta1 PodDisruptionBudget is deprecated in v1.21+, unavailable in v1.25+

### ISSUE THREAD

**Issue by pierluigilenoci** [kubernetes/autoscaler#4285]
**Which component are you using?**: cluster-autoscaler

**What version of the component are you using?**: 

Component version: v1.21.0, chart 9.10.5

**What k8s version are you using (`kubectl version`)?**:

<details><summary><code>kubectl version</code> Output</summary><br><pre>
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.1", GitCommit:"632ed300f2c34f6d6d15ca4cef3d3c7073412212", GitTreeState:"clean", BuildDate:"2021-08-19T15:38:26Z", GoVersion:"go1.16.6", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"21+", GitVersion:"v1.21.2-eks-0389ca3", GitCommit:"8a4e27b9d88142bbdd21b997b532eb6d493df6d2", GitTreeState:"clean", BuildDate:"2021-07-31T01:34:46Z", GoVersion:"go1.16.5", Compiler:"gc", Platform:"linux/amd64"}
</pre></details>

**What environment is this in?**:

I use Flux/Helm Operator to deploy the chart.

**What did you expect to happen?**:

I expect the chart to use, when possible, the latest versions of the APIs or to use the appropriate APIs based on the Kubernetes version of the cluster. 

**What happened instead?**:

```shell
W0826 12:53:50.687259       6 warnings.go:70] policy/v1beta1 PodDisruptionBudget is deprecated in v1.21+, unavailable in v1.25+; use policy/v1 PodDisruptionBudget
W0826 12:53:50.692111       6 warnings.go:70] policy/v1beta1 PodDisruptionBudget is deprecated in v1.21+, unavailable in v1.25+; use policy/v1 PodDisruptionBudget
W0826 12:53:50.712486       6 warnings.go:70] policy/v1beta1 PodDisruptionBudget is deprecated in v1.21+, unavailable in v1.25+; use policy/v1 PodDisruptionBudget
```

**How to reproduce it (as minimally and precisely as possible)**:

Install the helm chart in a cluster with K8s > 1.20 

**Anything else we need to know?**: 

Nothing

**Comment by pierluigilenoci** [comment]
@ipochi @bpineau @x13n @MaciekPytel could you please take a look?

**Comment by bpineau** [comment]
When cluster-autoscaler switchs its internal pdblister to use policy/v1 (instead of policy/v1beta1), we'll get rid of that warning but also the cluster-autoscaler will cease to work on any cluster older than v1.21 (they don't have policy/v1 at all).
Granted that's not an officially supported setup, but still noteworthy.

For ref, [the timeline for that resource ](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/) is:
* v1.3 - v1.4 : policy/v1alpha1 PodDisruptionBudget introduced
* v1.5 - v1.20: policy/v1beta1 PodDisruptionBudget
* v1.21 : policy/v1 PodDisruptionBudget introduced (GA) + policy/v1beta1 PodDisruptionBudget deprecated
* v1.25 : policy/v1beta1 PodDisruptionBudget removed

**Comment by pierluigilenoci** [comment]
@bpineau in the Chart you can use `Capabilities.APIVersions` to check which API is supported and use the correct version. Ref: https://helm.sh/docs/chart_template_guide/builtin_objects/ 

There are therefore no technical or backward compatibility reasons that prevent this issue from being resolved.

**Comment by pierluigilenoci** [comment]
For example, the guys from Bitnami use an interesting approach.
This https://github.com/bitnami/charts/blob/master/bitnami/common/templates/_capabilities.tpl#L65 
and then this https://github.com/bitnami/charts/blob/master/bitnami/wordpress/templates/ingress.yaml#L2

**Comment by bpineau** [comment]
Absolutely, and that can also be detected and implemented directly from the autoscaler code (so users not using that helm chart wouldn't have to deal with that, and autoscaler wouldn't need to introduce a cli flag that would be removed in 1.25).

I didn't mean to say there's a technical obstacle, but wanted to point out we should be careful and that's a choice to make (either abandoning old k8s, or supporting older versions and having a special case here - from a quick look around, it seems cluster-autoscaler has no other conditionals around api endpoints selection currently).

Meanwhile, the warnings are innocuous (no pdb schema change in beta->GA, one small semantic improvement (empty selector can now be used to selects all pods in the namespace)).

**Comment by pierluigilenoci** [comment]
@ipochi @bpineau @x13n @MaciekPytel @jbartosik  any news about this?

**Comment by x13n** [comment]
I think at this point it should be safe to just move to v1 in the code. CA 1.23 is out, so the changes will be released in 1.24 version. Things will start breaking if someone deploys CA 1.24 in 1.20 k8s cluster, which is a huge version skew that shouldn't be used anyway. Given CA dependency on scheduler code, having even one minor version skew isn't the best idea as it may lead to disagreement between scheduler and CA.

**Comment by pierluigilenoci** [comment]
@ipochi @bpineau @x13n @MaciekPytel @jbartosik any news about this?

**Comment by karlroberts** [comment]
could you not simply supply the v1 version if kubernetes semver >= 1.21  in the same way you do with deployment.apiVersion in _helpers.tpl?
![Screenshot from 2022-05-06 13-57-55](https://user-images.githubusercontent.com/462160/167064633-9b1feedd-602b-4fdc-961c-0fdf9327a5e5.png)

**Comment by x13n** [comment]
#4888 fixed the helm charts, but CA is still relying on v1beta1 internally, which has to be fixed before 1.25 release.

/reopen

**Comment by liggitt** [comment]
> which has to be fixed before 1.25 release.

for multi-server clusters, it actually has to be fixed in the 1.24 timeframe to work properly against a cluster that is upgrading from 1.24 to 1.25, where some of the API servers will not have the beta API available

**Comment by x13n** [comment]
Ah, good point. We'll need to cherrypick the change to 1.24 branch.
